Replace some sleeps with condition variables

dr-m opened the pull request:
I was hoping that this would allow the plugin to work with a smaller GPU memory. It actually did start up with `gpu_mem=128`, displaying a HD channel, which I think previously only worked with `gpu_mem=192`. Still, after pressing the Menu button 4 times (to show and hide full-screen OSD twice) it was unable to open any more full-screen OSD menu at `gpu_mem=128`.

On a quick test, the OSD still seems to be fine.

dr-m committed:
Fix some races and prefer memmove(), memset() to for loops

cOmx::PollVideo(), cOmx::GetBufferUsage(): Protect the calculation
with the mutex, and simplify the arithmetics.

dr-m committed:
Prefer memset() to a loop

dr-m commented:
Replacing the remaining `cCondWait::SleepMs()` would require large changes that may be impractical.

reufer commented on omx.c:
```diff
@@ -180,91 +119,92 @@ const char* cOmx::errStr(int err)
 void cOmx::HandlePortBufferEmptied(eOmxComponent component)
 {
-	Lock();
+	m_mutex.Lock();
```
why not using cThread's mutex?

dr-m replied:
`m_usedAudioBuffers[]` and `m_usedVideoBuffers[]` are protected by `cOmx::m_mutex` elsewhere. Sure, they could be protected by `cThread::mutex`, but then some code would have to acquire two mutexes. For these simple counters, it could make sense to upgrade to C++11 (which is supported starting with GCC 4.8.5), make the counters `std::atomic<int>` and use `fetch_sub(1)`, to avoid any race condition.

Unfortunately, `cThread::mutex` is not declared `protected`, but `private`. Therefore, there is no way for derived classes such as `cOmx` to use it together with a `cCondVar` (such as `cOmx::m_portEventsAdded`). To be able to use `cCondVar`, we are forced to declare a separate `cMutex` on our own.

Another somewhat unfortunate abstraction is that there is no variant of `cCondVar::TimedWait()` that would allow a previously set expiration period to be reused. In `cOmx::Action()`, we could replace the local variable `timer` with more clever logic: Set a `struct timespec abstime` 100 milliseconds into the future, and keep invoking `pthread_cond_timedwait()` with that expiration time. If and only if a timeout expires, set the timeout 100 milliseconds into the future and perform the periodic action. This would save quite a few context switches due to system calls and wakeups.

reufer replied:
Ok, thanks for the explanations. For the extension of `cCondVar::TimedWait()` it could make sense to submit a patch to Klaus, as the author of VDR, he's very open to sensible improvements of VDR's core.

dr-m replied:
Sure. I will first have to think about an API change that would allow `cOmx::Action()` to be rewritten in the way I envisioned.

dr-m replied:
I proposed some API changes in https://www.linuxtv.org/pipermail/vdr/2023-January/029700.html
and wrote a patch that makes use of it:
```diff
diff --git a/omx.c b/omx.c
index 7e517ec..07a6cb1 100644
--- a/omx.c
+++ b/omx.c
@@ -118,15 +118,26 @@ const char* cOmx::errStr(int err)
 
 void cOmx::Action(void)
 {
-	cTimeMs timer;
+	timespec abstime;
+	if (!cCondVar::GetAbsTime(&abstime, 100))
+		return;
 	m_mutex.Lock();
 	while (Running())
 	{
-		while (m_portEvents.empty())
-			if (m_portEventsAdded.TimedWait(m_mutex, 10))
-				goto timeout;
-
+		if (!m_portEvents.empty());
+		else if (!m_portEventsAdded.TimedWait(m_mutex, abstime))
 		{
+			memmove(m_usedAudioBuffers, m_usedAudioBuffers + 1,
+				(sizeof m_usedAudioBuffers) -
+				sizeof *m_usedAudioBuffers);
+			memmove(m_usedVideoBuffers, m_usedVideoBuffers + 1,
+				(sizeof m_usedVideoBuffers) -
+				sizeof *m_usedVideoBuffers);
+			if (!cCondVar::GetAbsTime(&abstime, 100))
+				break;
+			continue;
+		}
+
 		const Event event = m_portEvents.front();
 		m_portEvents.pop();
 		m_mutex.Unlock();
@@ -163,19 +174,6 @@ void cOmx::Action(void)
 
 		m_mutex.Lock();
 	}
-
-timeout:
-		if (timer.TimedOut())
-		{
-			timer.Set(100);
-			memmove(m_usedAudioBuffers, m_usedAudioBuffers + 1,
-				(sizeof m_usedAudioBuffers) -
-				sizeof *m_usedAudioBuffers);
-			memmove(m_usedVideoBuffers, m_usedVideoBuffers + 1,
-				(sizeof m_usedVideoBuffers) -
-				sizeof *m_usedVideoBuffers);
-		}
-	}
 	m_mutex.Unlock();
 }
 
```
The final patch would have to be conditional on the VDR version number, once something like my proposal becomes available.

My attempts to replace all `m_mutex` with `mutex` (e.g., replace `m_mutex.Lock()` with `Lock()`) led to deadlocks, as I noted in #15.

dr-m replied:
The call to `gettimeofday()` inside `GetAbsTime()` should be relatively cheap nowadays, thanks to the virtual system call implemented in `linux-vdso.so`, but with the above patch, the `abstime` could be advanced by more than 0.1 seconds in each step.

To avoid cumulative error, we should probably just add 0.1 seconds to the `abstime`. In case the system clock could be adjusted backwards, it could be a good idea to first ensure that `abstime` is at most a few seconds behind the current `gettimeofday()`.

dr-m replied:
The above idea is included in #16.

dr-m replied:
#15 answers the main question: Why can’t we use `cThread::mutex` a.k.a. `cOmx::mutex`? It protects "entry" calls to `ilclient_` and `OMX_`. The `cOmx::m_mutex` protects data structures in code that is invoked by callback functions of those libraries. The "main thread" that is holding `cOmx::mutex` may be blocked, waiting for those callback functions to complete. I believe that for this reason, the `cOmx::m_portEvents` queue is necessary, and the callback functions must use a different mutex. While working on #15, I tried to eliminate the queue, in vain, until I realized why the deadlocks occur.

dr-m committed:
cOvgThread: Synchronous CreateSurface(), DestroySurface()

Also, replace some timed waits with cCondVar::Wait().

dr-m committed:
Fix race conditions on cRpiAudioDecoder::cParser::m_parsed

dr-m committed:
Cleanup: Remove cRpiAudioRender::m_mutex

This code is only invoked via cRpiAudioDecoder::m_render,
mainly from cRpiAudioDecoder::Action(). Only m_render->Flush()
will be invoked from cRpiAudioDecoder::DeInit(), which will first
ensure that cRpiAudioDecoder::Action() has terminated.

reufer commented on ovgosd.c:
```diff
@@ -886,8 +886,12 @@ class cOvgCmdDestroySurface : public cOvgCmd
 	}
+
+private:
+	cOvgRenderTarget *&m_targetRef;
```
what is `m_targetRef` used for?

dr-m replied:
The assignment `m_targetRef = NULL` in `cOvgCmdDestroySurface::Execute()` will be observed by `cOvgThread::DestroySurface()`:
```c++
	void DestroySurface(cOvgRenderTarget *&buffer)
	{
		DoCmd(new cOvgCmdDestroySurface(buffer));
		m_commandsMutex.Lock();
		while (buffer)
			m_commandsLoop.Wait(m_commandsMutex);
		m_commandsMutex.Unlock();
	}
```

reufer replied:
ah, got it!

reufer commented on omx.h:
```diff
@@ -183,11 +201,11 @@ class cOmx : public cThread
-	static void OnEndOfStream(void *instance, COMPONENT_T *comp, OMX_U32 data);
-	static void OnError(void *instance, COMPONENT_T *comp, OMX_U32 data);
-	static void OnConfigChanged(void *instance, COMPONENT_T *comp, OMX_U32 data);
+	static void OnBufferEmpty(void *This, COMPONENT_T *comp);
```
what's wrong with `instance`?

dr-m replied:
There was nothing really wrong with that, but I thought that `This` better conveys the idea that the parameter is a type-punned pointer to a `cOmx` object, that is, `this`.

reufer replied:
I see - however, `This` looks a bit like misbehaved auto correction to me. What's about `void *omx`?

dr-m replied:
I can apply that, or I can retain the name `instance`. Which one would you prefer?

reufer replied:
then let's keep it an `instance`.

dr-m replied:
`instance` is preserved in 37a018d3d4d14947f44e68c7d8b7424853b369c3

reufer reviewed (commented):
Thanks a lot for the patches - just some minor questions to clarify.

dr-m committed:
Clean up cOmx::Event

Let us avoid some pointer indirection and heap allocation around
Event objects. Passing 2 words by value should not be worse than
allocating the data from heap.

dr-m committed:
Remove pointer indirection for cRpiSetup::s_instance

dr-m committed:
cRpiOsdProvider: Replace s_instance, m_ovg with s_ovg

dr-m committed:
Remove pointer indirection for cOvgFont::s_fonts

reufer commented on omx.h:
```diff
@@ -186,7 +204,7 @@ class cOmx : public cThread
 	static void OnPortSettingsChanged(void *instance, COMPONENT_T *comp, OMX_U32 data);
 	static void OnEndOfStream(void *instance, COMPONENT_T *comp, OMX_U32 data);
-	static void OnError(void *instance, COMPONENT_T *comp, OMX_U32 data);
+	static void OnError(void *, COMPONENT_T *, OMX_U32 data);
```
I think this change was not intentionally?

dr-m replied:
That was intentional, to document that the two parameters are ignored by the function.

reufer replied:
ok, makes sense.
